pkg/restore/restore.go

```diff
@@ -533,6 +677,107 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 	return warnings, errs
 }
 
+func (ctx *context) restoreVolumeFromSnapshot(obj *unstructured.Unstructured) (*unstructured.Unstructured, error, error) {
+	spec, err := collections.GetMap(obj.UnstructuredContent(), "spec")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// if it's an unsupported volume type for snapshot restores, don't try to
+	// do a snapshot restore
+	if sourceType, _ := kube.GetPVSource(spec); sourceType == "" {
+		return obj, nil, nil
+	}
+
+	var (
+		pvName              = obj.GetName()
+		restoreFromSnapshot = false
+		restore             = ctx.restore
+		backup              = ctx.backup
+	)
+
+	if restore.Spec.RestorePVs != nil && *restore.Spec.RestorePVs {
+		// when RestorePVs = yes, it's an error if we don't have a snapshot service
+		if ctx.snapshotService == nil {
+			return nil, nil, errors.New("PV restorer is not configured for PV snapshot restores")
+		}
+
+		// if there are no snapshots in the backup, return without error
+		if backup.Status.VolumeBackups == nil {
+			return obj, nil, nil
+		}
+
+		// if there are snapshots, and this is a supported PV type, but there's no
+		// snapshot for this PV, it's an error
+		if backup.Status.VolumeBackups[pvName] == nil {
+			return nil, nil, errors.Errorf("no snapshot found to restore volume %s from", pvName)
+		}
+
+		restoreFromSnapshot = true
+	}
+	if restore.Spec.RestorePVs == nil && ctx.snapshotService != nil {
+		// when RestorePVs = Auto, don't error if the backup doesn't have snapshots
+		if backup.Status.VolumeBackups == nil || backup.Status.VolumeBackups[pvName] == nil {
+			return obj, nil, nil
+		}
+
+		restoreFromSnapshot = true
+	}
+
+	if restoreFromSnapshot {
+		backupInfo := backup.Status.VolumeBackups[pvName]
+
+		ctx.infof("restoring PersistentVolume %s from SnapshotID %s", pvName, backupInfo.SnapshotID)
+		volumeID, err := ctx.snapshotService.CreateVolumeFromSnapshot(backupInfo.SnapshotID, backupInfo.Type, backupInfo.AvailabilityZone, backupInfo.Iops)
+		if err != nil {
+			return nil, nil, err
+		}
+		ctx.infof("successfully restored PersistentVolume %s from snapshot", pvName)
+
+		if err := kube.SetVolumeID(spec, volumeID); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	var warning error
+
+	if ctx.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
+		warning = errors.New("unable to restore PV snapshots: Ark server is not configured with a PersistentVolumeProvider")
+	}
+
+	return obj, warning, nil
```

Comment:
Is having 2 `error` fields on return with one for warnings a Go idiom? I think I understand that it's meant to protect against callers being overzealous with `if err != nil`, but it looks like there's only one case here where we raise a warning in this function

Reply:
This is because we (ark) support both warnings and errors when restoring. This allows individual restore actions to distinguish as well.